File: picocapsule.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""picocapsule.py -- capsule manifest model + deterministic canonical-text
serializer/parser, and pack/card address helpers.

This is the PicoScript reference implementation of the capsule contract handed
to the PIOS build agent in docs/PIOS_CAPSULE_HANDOFF.md.  A capsule is a pack
namespace; card 0 holds the manifest in the frozen canonical text format (see
that doc, section 3).  serialize() is fully deterministic so card 0 is byte
stable, and parse(serialize(m)) == m for any manifest.

The browser mirror (vm/picocapsule.js) must emit byte-identical text so a
manifest authored in either runtime produces the same card 0.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import List, Optional

# Capsule pack-id range (defaults; a deployment may widen these -- the manifest
# is authoritative, see the handoff doc).
CAPSULE_PACK_MIN = 1024
CAPSULE_PACK_MAX = 4095

# Default source/bytecode card pairing convention: program N.
SOURCE_BASE = 1000
CODE_BASE = 10000

_NAME_RE = re.compile(r"^[A-Za-z0-9_-]+$")


def source_for(n: int) -> int:
    """Default source card for program N (1000 + N)."""
    return SOURCE_BASE + n


def code_for(n: int) -> int:
    """Default bytecode card for program N (10000 + N)."""
    return CODE_BASE + n


# ── pack/card addressing ─────────────────────────────────────────────────────

def format_address(pack: int, card: int) -> str:
    """Canonical address: 'pack/card' (e.g. '1024/10001')."""
    return f"{pack}/{card}"


def parse_address(text: str) -> tuple:
    """Parse 'pack/card' or typed 'capsule:P/card:C' -> (pack, card)."""
    s = text.strip()
    left, sep, right = s.partition("/")
    if not sep:
        raise ValueError(f"bad address {text!r}: expected 'pack/card'")
    if ":" in left:                      # typed form: capsule:1024/card:10001
        pack = int(left.split(":", 1)[1])
        card = int(right.split(":", 1)[1])
    else:
        pack, card = int(left), int(right)
    return pack, card


def is_capsule_pack(pack: int) -> bool:
    return CAPSULE_PACK_MIN <= pack <= CAPSULE_PACK_MAX


# ── manifest model ───────────────────────────────────────────────────────────

@dataclass
class Process:
    name: str
    source: int
    bytecode: int
    io: Optional[str] = None        # "tcp/83" | "fifo/requests"
    entry: Optional[str] = None     # "http", ...


@dataclass
class Fifo:
    name: str
    frm: str
    to: str
    depth: int
    frame_max: int


@dataclass
class Manifest:
    name: str
    cards: str = "1001-20000"       # "<lo>-<hi>"
    principal: Optional[str] = None
    mem_kib: Optional[int] = None
    cpu_ms: Optional[int] = None
    fs: Optional[str] = None
    processes: List[Process] = field(default_factory=list)
    fifos: List[Fifo] = field(default_factory=list)

    # builder helpers (mirror the Capsule.* / Manifest.* primitives) -----------
    def process(self, name: str, source: int, bytecode: int,
                io: Optional[str] = None, entry: Optional[str] = None) -> "Manifest":
        self.processes.append(Process(name, source, bytecode, io, entry))
        return self

    def bind_tcp(self, port: int, process: str) -> "Manifest":
        for p in self.processes:
            if p.name == process:
                p.io = f"tcp/{port}"
                return self
        raise ValueError(f"bind: no process named {process!r}")

    def fifo(self, name: str, frm: str, to: str, depth: int, frame_max: int) -> "Manifest":
        self.fifos.append(Fifo(name, frm, to, depth, frame_max))
        return self


# ── deterministic canonical-text serializer (handoff doc section 3) ──────────

def _check_name(kind: str, name: str) -> None:
    if not _NAME_RE.match(name):
        raise ValueError(f"{kind} name {name!r} must match [A-Za-z0-9_-]+")


def serialize(m: Manifest) -> str:
    """Emit the frozen canonical manifest text (UTF-8, '\\n' endings).

    Header key order: capsule, name, [principal, mem_kib, cpu_ms, fs], cards.
    Then process blocks (declaration order), then ipc_fifo blocks. Trailing \\n.
    """
    _check_name("capsule", m.name)
    out: List[str] = ["capsule = on", f"name = {m.name}"]
    if m.principal is not None:
        out.append(f"principal = {m.principal}")
    if m.mem_kib is not None:
        out.append(f"mem_kib = {m.mem_kib}")
    if m.cpu_ms is not None:
        out.append(f"cpu_ms = {m.cpu_ms}")
    if m.fs is not None:
        out.append(f"fs = {m.fs}")
    out.append(f"cards = {m.cards}")
    for p in m.processes:
        _check_name("process", p.name)
        out.append("")
        out.append(f"process = {p.name}")
        out.append(f"  source = {p.source}")
        out.append(f"  bytecode = {p.bytecode}")
        if p.io is not None:
            out.append(f"  io = {p.io}")
        if p.entry is not None:
            out.append(f"  entry = {p.entry}")
    for fo in m.fifos:
        _check_name("ipc_fifo", fo.name)
        out.append("")
        out.append(f"ipc_fifo = {fo.name}")
        out.append(f"  from = {fo.frm}")
        out.append(f"  to = {fo.to}")
        out.append(f"  depth = {fo.depth}")
        out.append(f"  frame_max = {fo.frame_max}")
    return "\n".join(out) + "\n"


def parse(text: str) -> Manifest:
    """Parse canonical manifest text back into a Manifest (inverse of serialize)."""
    m = Manifest(name="")
    cur = None
    for raw in text.split("\n"):
        line = raw.rstrip("\r")
        if line.strip() == "":
            continue
        indented = line.startswith("  ")
        key, sep, val = line.strip().partition("=")
        if not sep:
            raise ValueError(f"bad manifest line {line!r}: expected 'key = value'")
        key, val = key.strip(), val.strip()
        if not indented:
            if key == "capsule":
                continue
            elif key == "name":
                m.name = val
            elif key == "principal":
                m.principal = val
            elif key == "mem_kib":
                m.mem_kib = int(val)
            elif key == "cpu_ms":
                m.cpu_ms = int(val)
            elif key == "fs":
                m.fs = val
            elif key == "cards":
                m.cards = val
            elif key == "process":
                cur = Process(val, 0, 0)
                m.processes.append(cur)
            elif key == "ipc_fifo":
                cur = Fifo(val, "", "", 0, 0)
                m.fifos.append(cur)
            else:
                raise ValueError(f"unknown manifest key {key!r}")
        else:
            if isinstance(cur, Process):
                if key == "source":
                    cur.source = int(val)
                elif key == "bytecode":
                    cur.bytecode = int(val)
                elif key == "io":
                    cur.io = val
                elif key == "entry":
                    cur.entry = val
                else:
                    raise ValueError(f"unknown process key {key!r}")
            elif isinstance(cur, Fifo):
                if key == "from":
                    cur.frm = val
                elif key == "to":
                    cur.to = val
                elif key == "depth":
                    cur.depth = int(val)
                elif key == "frame_max":
                    cur.frame_max = int(val)
                else:
                    raise ValueError(f"unknown ipc_fifo key {key!r}")
            else:
                raise ValueError(f"indented line {line!r} outside a block")
    return m

File: vm/picocapsule.js
// picocapsule.js -- capsule manifest model + deterministic canonical-text
// serializer/parser + pack/card address helpers, for the browser/editor.
//
// Byte-identical to picocapsule.py: serialize() must emit exactly the bytes that
// land in capsule card 0 regardless of which runtime authored the manifest. See
// docs/PIOS_CAPSULE_HANDOFF.md section 3 (the frozen canonical format).
(function (root, factory) {
  var P = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = P;
  else root.PicoCapsule = P;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  "use strict";

  // Capsule pack-id range (defaults; the manifest is authoritative).
  var CAPSULE_PACK_MIN = 1024, CAPSULE_PACK_MAX = 4095;
  // Default source/bytecode card pairing for program N.
  var SOURCE_BASE = 1000, CODE_BASE = 10000;
  var NAME_RE = /^[A-Za-z0-9_-]+$/;

  function sourceFor(n) { return SOURCE_BASE + n; }   // program N source card
  function codeFor(n) { return CODE_BASE + n; }       // program N bytecode card

  // ---- pack/card addressing -------------------------------------------------
  function formatAddress(pack, card) { return pack + "/" + card; }
  function parseAddress(text) {
    var s = ("" + text).trim();
    var slash = s.indexOf("/");
    if (slash < 0) throw new Error("bad address " + JSON.stringify(text) + ": expected 'pack/card'");
    var left = s.slice(0, slash), right = s.slice(slash + 1);
    var pack, card;
    if (left.indexOf(":") >= 0) {           // typed: capsule:1024/card:10001
      pack = parseInt(left.split(":")[1], 10);
      card = parseInt(right.split(":")[1], 10);
    } else {
      pack = parseInt(left, 10); card = parseInt(right, 10);
    }
    return [pack, card];
  }
  function isCapsulePack(pack) { return pack >= CAPSULE_PACK_MIN && pack <= CAPSULE_PACK_MAX; }

  // ---- manifest model -------------------------------------------------------
  function Manifest(name, cards) {
    this.name = name;
    this.cards = cards || "1001-20000";
    this.principal = null; this.mem_kib = null; this.cpu_ms = null; this.fs = null;
    this.processes = []; this.fifos = [];
  }
  Manifest.prototype.process = function (name, source, bytecode, io, entry) {
    this.processes.push({ name: name, source: source, bytecode: bytecode,
                          io: (io === undefined ? null : io), entry: (entry === undefined ? null : entry) });
    return this;
  };
  Manifest.prototype.bindTcp = function (port, process) {
    for (var i = 0; i < this.processes.length; i++) {
      if (this.processes[i].name === process) { this.processes[i].io = "tcp/" + port; return this; }
    }
    throw new Error("bind: no process named " + process);
  };
  Manifest.prototype.fifo = function (name, frm, to, depth, frameMax) {
    this.fifos.push({ name: name, frm: frm, to: to, depth: depth, frame_max: frameMax });
    return this;
  };

  function checkName(kind, name) {
    if (!NAME_RE.test(name)) throw new Error(kind + " name " + JSON.stringify(name) + " must match [A-Za-z0-9_-]+");
  }

  // ---- deterministic canonical-text serializer (handoff doc section 3) ------
  function serialize(m) {
    checkName("capsule", m.name);
    var out = ["capsule = on", "name = " + m.name];
    if (m.principal != null) out.push("principal = " + m.principal);
    if (m.mem_kib != null) out.push("mem_kib = " + m.mem_kib);
    if (m.cpu_ms != null) out.push("cpu_ms = " + m.cpu_ms);
    if (m.fs != null) out.push("fs = " + m.fs);
    out.push("cards = " + m.cards);
    for (var i = 0; i < m.processes.length; i++) {
      var p = m.processes[i]; checkName("process", p.name);
      out.push(""); out.push("process = " + p.name);
      out.push("  source = " + p.source);
      out.push("  bytecode = " + p.bytecode);
      if (p.io != null) out.push("  io = " + p.io);
      if (p.entry != null) out.push("  entry = " + p.entry);
    }
    for (var j = 0; j < m.fifos.length; j++) {
      var f = m.fifos[j]; checkName("ipc_fifo", f.name);
      out.push(""); out.push("ipc_fifo = " + f.name);
      out.push("  from = " + f.frm);
      out.push("  to = " + f.to);
      out.push("  depth = " + f.depth);
      out.push("  frame_max = " + f.frame_max);
    }
    return out.join("\n") + "\n";
  }

  function parse(text) {
    var m = new Manifest("");
    var cur = null, curType = null;
    var lines = text.split("\n");
    for (var i = 0; i < lines.length; i++) {
      var line = lines[i].replace(/\r$/, "");
      if (line.trim() === "") continue;
      var indented = line.slice(0, 2) === "  ";
      var eq = line.indexOf("=");
      if (eq < 0) throw new Error("bad manifest line " + JSON.stringify(line) + ": expected 'key = value'");
      var key = line.slice(0, eq).trim(), val = line.slice(eq + 1).trim();
      if (!indented) {
        if (key === "capsule") continue;
        else if (key === "name") m.name = val;
        else if (key === "principal") m.principal = val;
        else if (key === "mem_kib") m.mem_kib = parseInt(val, 10);
        else if (key === "cpu_ms") m.cpu_ms = parseInt(val, 10);
        else if (key === "fs") m.fs = val;
        else if (key === "cards") m.cards = val;
        else if (key === "process") { cur = { name: val, source: 0, bytecode: 0, io: null, entry: null }; m.processes.push(cur); curType = "process"; }
        else if (key === "ipc_fifo") { cur = { name: val, frm: "", to: "", depth: 0, frame_max: 0 }; m.fifos.push(cur); curType = "fifo"; }
        else throw new Error("unknown manifest key " + key);
      } else if (curType === "process") {
        if (key === "source") cur.source = parseInt(val, 10);
        else if (key === "bytecode") cur.bytecode = parseInt(val, 10);
        else if (key === "io") cur.io = val;
        else if (key === "entry") cur.entry = val;
        else throw new Error("unknown process key " + key);
      } else if (curType === "fifo") {
        if (key === "from") cur.frm = val;
        else if (key === "to") cur.to = val;
        else if (key === "depth") cur.depth = parseInt(val, 10);
        else if (key === "frame_max") cur.frame_max = parseInt(val, 10);
        else throw new Error("unknown ipc_fifo key " + key);
      } else {
        throw new Error("indented line " + JSON.stringify(line) + " outside a block");
      }
    }
    return m;
  }

  return {
    Manifest: Manifest, serialize: serialize, parse: parse,
    formatAddress: formatAddress, parseAddress: parseAddress,
    sourceFor: sourceFor, codeFor: codeFor, isCapsulePack: isCapsulePack,
    CAPSULE_PACK_MIN: CAPSULE_PACK_MIN, CAPSULE_PACK_MAX: CAPSULE_PACK_MAX
  };
});
